### PR TITLE
[FIX] core: reintroduce model dependencies

### PR DESCRIPTION
**Pull request by rco-odoo**
Make sure that model dependencies are flushed before searching with a
given domain.  This fixes inconsistencies in `search` and `read_group`.

### CODE REVIEW

**Comment by rco-odoo** [comment]
@william-andre can you validate that it fixes your issue?

**Comment by william-andre** [comment]
@rco-odoo Yes, tested here https://github.com/odoo/odoo/commit/a3d00ff9aec5620964263d2e23de812f892444bc#diff-736254d636e6f1819d52423bed609ccbR237-R240, runbot is green :+1:

**Comment by rco-odoo** [comment]
@robodoo rebase-ff r+

File: odoo/models.py [hunk 1]
```diff
@@ -5391,9 +5391,11 @@ def update(self, values):
...
     @api.model
     def flush(self, fnames=None, records=None):
-        """ Process all the pending recomputations (or at least the given field
-            names `fnames` if present) and flush the pending updates to the
-            database.
+        """ When called without parameters, process all the pending computations
+        on all models, and flush the pending updates to the database.
+        If ``fnames`` is given, the processing is limited to the given fields of
+        the current model, and even to the given records (if both ``fnames`` and
+        ``records`` are given).
```
Comment (william-andre) [inline odoo/models.py:5405]: It is maybe not the moment/place to discuss docstring guidelines, but I find the PEP257 way easier to read
```python
    def flush(self, fnames=None, records=None):
        """Process the all the pending computations  and flush to the database.
        
        :param fnames (list<string>): list of field names to be flushed. 
            If it is given, the processing is limited to the given fields
            of the current model.
        :param records (Model): the recordset that needs to be processed. 
            It has no effect if ``fnames`` is not given
        """
```

**Comment by rco-odoo** [comment]
@robodoo r-

**Comment by rco-odoo** [comment]
@robodoo r+

**Review by odony — COMMENTED** [review]
Thanks, I guess that's a quick fix with no measurable performance impact for the normal models.

I had a quick look at the `_depends` definition we still have around in 13.0 (crm + accounting and l10n), to make sure they won't suddenly backfire, and I did not notice any invalid model or field name in there. The `_depends` are not exhaustive, but that's another story.

With regard to modularity I have 2 questions:
- is the flush mechanism working recursively for views, i.e. when a view depends on another view (we have that case with [crm.partner.report.assign](https://github.com/odoo/odoo/blob/13.0/addons/website_crm_partner_assign/report/crm_partner_report.py#L27)
- is the extension mechanism for extending `_depends` through `_inherit` still working? We [still have that use case](https://github.com/odoo/odoo/blob/13.0/addons/l10n_latam_invoice_document/report/invoice_report.py#L10) too

**Comment by rco-odoo** [comment]
@robodoo r-

**Comment by rco-odoo** [comment]
> With regard to modularity I have 2 questions:
> 
> * is the flush mechanism working recursively for views, i.e. when a view depends on another view (we have that case with [crm.partner.report.assign](https://github.com/odoo/odoo/blob/13.0/addons/website_crm_partner_assign/report/crm_partner_report.py#L27)

Very good question.  I slightly modified the code to make the collecting of dependencies recursive.

> * is the extension mechanism for extending `_depends` through `_inherit` still working? We [still have that use case](https://github.com/odoo/odoo/blob/13.0/addons/l10n_latam_invoice_document/report/invoice_report.py#L10) too

Oops, I forgot that part.  Thanks for pointing it out!

File: odoo/models.py [hunk 2]
```diff
@@ -336,6 +336,11 @@ class BaseModel(MetaModel('DummyModel', (object,), {'_register': False})):
     as attribute.
     """
 
+    _depends = {}
+    """dependencies of models backed up by SQL views
+    ``{model_name: field_names}``, where ``field_names`` is any iterable
```
Comment (odony) [inline odoo/models.py:341]: Perhaps we can help developers understand what to expect with an extra sentence like:

_This is only used to determine the changes to flush to database before executing `search()` or `read_group()`. It won't be used for cache invalidation or recomputing fields_

**Review by odony — APPROVED** [review]
That works for me, thanks! :+1: 

Ideally we should have some basic tests for the flushing mechanism, but I don't mind adding them in a follow-up commit, in order to provide an immediate solution for the bug @william-andre is trying to solve in 13.3

On a related note, I suggest targeting 13.3 as the first step, where we can battle-test it a bit (it impacts core stuff), and then backporting it to 13.0 after we confirm everything is ok. The backport PR could be prepared right now of course.


---

For the record, some rationale for the partial `_depends` support (possibly useful for the commit message):

- We choose to support `_depends`, in order to provide a mechanism for auto-flushing the dependencies of SQL views before they are queried by `search()` or `read_group()`. It turns out this is a common use case, and currently in 13.0 all callsites have to explicitly remember that views have no dependencies and have to know *what* to flush before querying them. It's not only error-prone, but not done in most cases, and we rely on luck and accidental flushes instead.
- We notice that developers tend to add flush() calls in unit tests, because they don't really know where the root problem lies, so in reality they hide it.
- On alternative would be to override `_flush_search()` on each view model, and manually call `flush()` as needed. But this is not easily extensible (duplicate flush() calls), and it's tricky to do correctly and efficiently. If done incorrectly, it can lead to performance issues when a view model is repeatedly queries in a transaction (e.g. for computing fields)
- We however decide *not* to support `_depends` for _cache invalidation_ and _recomputes_ because:
   1. It would have a more significant impact on ORM performance, considering those views often have lots of dependencies in commonly used models. The auto-flushing, on the other hand, only needs to trigger when *those SQL views* are queried.
   2. It would be useless most of the time, as those view records are almost never put in cache via `browse()` or `read()` individually. The real underlying records are used instead for anything that is not related to counting or aggregating.

**Comment by william-andre** [comment]
>   We however decide _not_ to support `_depends` for _cache invalidation_ and _recomputes_ because

Doesn't `flush` call recompute?

> We notice that developers tend to add flush() calls in unit tests, because they don't really know where the root problem lies, so in reality they hide it.

Just for the record, I added it in the test because in practice, the `read_group` on the analysis report is done in another transaction; and therefore everything is already flushed.

**Comment by odony** [comment]
> > We however decide _not_ to support `_depends` for _cache invalidation_ and _recomputes_ because
> 
> Doesn't `flush` call recompute?

Hmm, what I meant is that you can't `@depend` on view fields because they are not correctly invalidated in cache, so your field won't be recomputed - you need to depend on the "real" underlying fields, even if you have defined `_depends`. But that may deserve clarification ;-)


> Just for the record, I added it in the test because in practice, the `read_group` on the analysis report is done in another transaction; and therefore everything is already flushed.

Sure, in your case, though in practice might naturally have called `commit()`, not `flush()`, in order to simulate transactions.
But other scenarios involve these things successively in the same transaction or test: start by executing a business flow, then query a view to get some aggregated numbers based on the result, and update some other thing elsewhere with that info. With that scenario the developer is forced to manually call `flush()`, unless it works by chance due to a lucky flush elsewhere (= a lurking bug).
And all those cases are:
- *unexpected*: because the responsibility of flushing lies on the developer who queries the view even though they're not the one who designed the underlying dependencies, and they did not write any SQL that would justify it
- *error-prone*: because the missing flush() is tricky to spot, and may be partially hidden by other unrelated flush() calls
- *error-prone*: because doing the correct surgical flush() is non-trivial and verbose. And calling a global flush() is brutal and can be a real perf killer at the wrong spot

**Comment by rco-odoo** [comment]
@fw-bot ignore
@robodoo r+

**Comment by rco-odoo** [comment]
@robodoo rebase-ff

**Pull request closed without merging**